akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1148,6 +1159,20 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
       }
       internalInvoke(event)
     }
+
+    // external call
+    def cancel(): Unit = {
```

Comment:
final tailrec?

Reply:
Made entire class final up above, but should tailrec.